stats/json/collector.go

```diff
@@ -34,20 +37,18 @@ import (
 )
 
 type Collector struct {
-	outfile     io.WriteCloser
+	closeFn     func() error
 	fname       string
 	seenMetrics []string
-}
 
-// Verify that Collector implements lib.Collector
-var _ lib.Collector = &Collector{}
+	encoder *json.Encoder
 
-// Similar to ioutil.NopCloser, but for writers
-type nopCloser struct {
-	io.Writer
+	buffer     []stats.Sample
+	bufferLock sync.Mutex
```

Comment:
I suggest to change this `bufferLock` to something like:

```
    mu     sync.Mutex       // guard buffer
    buffer []stats.Sample
```

`bufferLock`, while its name indicates that it provides lock for buffer, but actually, it can provide lock for the `Collector`. Also using `c.bufferLock.Lock()` is something not idiom.

Reply:
hmm but looking at how this is used in the code below:
```go
func (c *Collector) Collect(scs []stats.SampleContainer) {
	c.bufferLock.Lock()
	defer c.bufferLock.Unlock()
	for _, sc := range scs {
		c.buffer = append(c.buffer, sc.GetSamples()...)
	}
}
```
to me that's much clearer than if we rename `bufferLock` to `mu`...
```go
func (c *Collector) Collect(scs []stats.SampleContainer) {
	c.mu.Lock()
	defer c.mu.Unlock()
	for _, sc := range scs {
		c.buffer = append(c.buffer, sc.GetSamples()...)
	}
}
```

So, while `bufferLock` might not be the best name, IMO mostly because of the repetition/stuttering when you have `bufferLock.Lock()`, `mu` is definitely a downgrade... an improvement would be something even more descriptive like  `bufferMutex` or something like that, but even that's a minor one...

> but actually, it can provide lock for the Collector

but this is actually false, it this mutex only applies to the `buffer`, anything else in the collector isn't affected.

Reply:
I am also of the opinion that it will not help the readability in any way and if anything will make it look worse as it will look as if there is mutex for the whole collector ... and then we aren't using for the whole collector